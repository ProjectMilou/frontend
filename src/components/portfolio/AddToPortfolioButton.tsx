import React from 'react';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import Checkbox from '@material-ui/core/Checkbox';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import MuiDialogTitle from '@material-ui/core/DialogTitle';
import MuiDialogContent from '@material-ui/core/DialogContent';
import MuiDialogActions from '@material-ui/core/DialogActions';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import { useTranslation } from 'react-i18next';
import * as API from '../../portfolio/APIClient';
import { errorMessageKey, errorTitleKey } from '../../Errors';
import ProgressButton from './ProgressButton';

// stylesheet for the entire add to portfolio component
const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    addButton: {
      padding: '0.25rem 1rem',
      backgroundColor: '#3fbcf2',
      '&:hover': {
        backgroundColor: '#84d4f7',
      },
      whiteSpace: 'nowrap',
    },
    Dialog: {
      minWidth: '40rem',
    },
    dialogHeader: {
      margin: 0,
      padding: theme.spacing(2),
      display: 'flex',
      justifyContent: 'center',
    },
    // dialog header
    headerCloseButton: {
      position: 'relative',
      color: theme.palette.grey[500],
    },
    span: {
      display: 'flex',
    },
    text: {
      margin: 'auto',
    },
    // dialog body
    DialogContent: {
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'space-between',
      margin: 0,
      padding: theme.spacing(1),
    },
    // List component
    List: {
      width: '100%',
      backgroundColor: 'white',
    },
    listItem: {
      width: '100%',
      margin: '0.75rem 0',
      height: '3.25rem',
    },
    // error helper text
    helperText: {
      position: 'absolute',
      top: '100%',
      right: '-50%',
      whiteSpace: 'nowrap',
    },
    dialogFooter: {
      margin: 0,
      padding: theme.spacing(2),
      display: 'flex',
      justifyContent: 'end',
    },
  })
);

type ListEntryProps = {
  id: string;
  name: string;
  checked: CheckedType;
  handleToggle: (id: string) => () => void;
  changeAmount: (id: string, amount: string) => void;
  pattern: RegExp;
};

const ListEntry: React.FC<ListEntryProps> = ({
  id,
  name,
  checked,
  handleToggle,
  changeAmount,
  pattern,
}) => {
  const classes = useStyles();

  return (
    <ListItem role={undefined} dense className={classes.listItem}>
      <ListItemIcon>
        <Checkbox
          edge="start"
          checked={typeof checked[id] !== 'undefined'}
          tabIndex={-1}
          disableRipple
          inputProps={{
            'aria-labelledby': `checkbox-list-label-${id}`,
          }}
          color="secondary"
          onClick={handleToggle(id)}
        />
      </ListItemIcon>
      <ListItemText primary={name} />
      <TextField
        variant="outlined"
        value={checked[id]}
        error={!pattern.test(checked[id])}
        helperText={
          pattern.test(checked[id]) ? '' : 'Please enter a positive number!'
        }
        FormHelperTextProps={{
          className: classes.helperText,
        }}
        onChange={(e) => {
          changeAmount(id, e.target.value);
        }}
        size="small"
        margin="dense"
        inputProps={{
          maxLength: 5,
          style: {
            textAlign: 'center',
            paddingRight: 0,
            paddingLeft: '0.2rem',
          },
        }}
        style={{
          display: typeof checked[id] !== 'undefined' ? 'block' : 'none',
          width: '4rem',
        }}
      />
    </ListItem>
  );
};

type CheckedType = {
  // id of portfolio: qty of stock;
  [id: string]: string;
};

type AddToPortfolioButtonType = {
  isin: string;
  token: string;
};

// returns the add to portfolio button and all subcomponents including the dialog window
const AddToPortfolioButton: React.FC<AddToPortfolioButtonType> = ({
  isin,
  token,
}) => {
  // style and translation hooks
  const classes = useStyles();
  const { t } = useTranslation();
  // pattern for testing input fields
  const pattern = /^(?:[1-9]\d*|0)?(?:\.\d+)?$/;

  // the list of portfolios belonging to the user
  const [portfolios, setPortfolios] = React.useState<API.PortfolioOverview[]>(
    []
  );

  // the state that controls whether the dialog window is open or closed
  const [open, setOpen] = React.useState(false);
  // true when waiting for an API response
  const [loading, setLoading] = React.useState<boolean>(false);
  const [error, setError] = React.useState<Error | undefined>();
  // keep track of which portfolios have been selected and the amount in the textfield
  const [checked, setChecked] = React.useState<CheckedType>({});

  // defines function to fetch from api
  const isMounted = React.useRef(true);
  const fetch = async () => {
    setError(undefined);
    try {
      const p = await API.list(token);
      if (isMounted.current) {
        setPortfolios(p);
      }
    } catch (e) {
      if (isMounted.current) {
        setError(e);
      }
    }
  };

  // opon mount fetch list of portfolios from api
  React.useEffect(() => {
    fetch();
    return () => {
      isMounted.current = false;
    };
    // deps must be empty because the function should only be called on mount.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const openDialog = () => {
    // reset the dialog's state on (re-)open
    setLoading(false);
    setError(undefined);
    setChecked({});
    setOpen(true);
  };

  // handle saving selections
  const handleSave = async () => {
    setLoading(true);

    try {
      // parse the checked state into the form expected by the api
      const modifications: API.PortfoliosQty[] = Object.entries(checked).map(
        ([id, qty]) => ({
          id,
          // TODO: use react-number-format for validation,
          //  store numbers instead of strings, avoid converting
          qty: Number.parseFloat(qty),
        })
      );
      // make api put call
      await API.stock(token, isin, modifications);
      setOpen(false);
    } catch (e) {
      // on error stop the loading animation and display error to user
      setLoading(false);
      setError(e);
    }
  };

  // handles checking and unchecking portfolios
  const handleCheckToggle = (id: string) => () => {
    const newChecked = { ...checked };

    if (typeof newChecked[id] === 'undefined') {
      // if unchecked, check it with an initial value of 1
      newChecked[id] = '1';
    } else {
      // if checked remove check
      delete newChecked[id];
    }

    setChecked(newChecked);
  };

  //  a function that sets the amount everytime a textfield is changed
  const changeAmount = (id: string, amount: string) => {
    // when a user changes the input field the checked state gets updated
    setChecked({ ...checked, [id]: amount });
  };

  return (
    <div>
      <Button
        variant="contained"
        className={classes.addButton}
        onClick={() => openDialog()}
      >
        {t('analyzer.addToPortfolio')}
      </Button>
      <Dialog
        onClose={() => setOpen(false)}
        aria-labelledby="dialog-title"
        open={open}
        className={classes.Dialog}
      >
        {/* The header of the dialog window */}
        <MuiDialogTitle disableTypography className={classes.dialogHeader}>
          <span className={classes.span}>
            <Typography variant="h6" className={classes.text}>
              {t('analyzer.dialogHeader')}
            </Typography>
          </span>
        </MuiDialogTitle>
        {/* The body of the dialog window */}
        <MuiDialogContent dividers className={classes.DialogContent}>
          <List className={classes.List} component="ol">
            {portfolios.map((portfolio) => (
              <ListEntry
                key={portfolio.id}
                id={portfolio.id}
                name={portfolio.name}
                checked={checked}
                handleToggle={handleCheckToggle}
                changeAmount={changeAmount}
                pattern={pattern}
              />
            ))}
          </List>
          {error && (
            <Typography color="error">
              <b>{t(errorTitleKey(error))}</b>: {t(errorMessageKey(error))}
            </Typography>
          )}
        </MuiDialogContent>
        {/* The footer of the dialog window */}
        <MuiDialogActions className={classes.dialogFooter}>
          <Button autoFocus onClick={() => setOpen(false)} color="primary">
            {t('portfolio.details.discardChanges')}
          </Button>
          <ProgressButton
            onClick={handleSave}
            color="primary"
            loading={loading}
            disabled={
              !Object.values(checked)
                .map((stringAmount) => pattern.test(stringAmount))
                .every((v) => v)
            }
          >
            {t('portfolio.details.saveChanges')}
          </ProgressButton>
        </MuiDialogActions>
      </Dialog>
    </div>
  );
};

// exports
export default AddToPortfolioButton;
