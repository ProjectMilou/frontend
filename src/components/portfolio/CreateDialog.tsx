import React from 'react';
import { Trans, useTranslation } from 'react-i18next';
import NameDialog from './NameDialog';

function validateName(newName?: string): boolean {
  const n = newName?.trim();
  return !!(n && n.length);
}

export type CreateDialogProps = {
  open: boolean;
  handleClose: () => void;
  create: (name: string) => Promise<void>;
};

const CreateDialog: React.FC<CreateDialogProps> = ({
  open,
  handleClose,
  create,
}) => {
  const { t } = useTranslation();

  return (
    <NameDialog
      open={open}
      dialogTitle={t('portfolio.dialog.create.title')}
      dialogText={<Trans i18nKey="portfolio.dialog.create.text" />}
      labelKey="portfolio.name"
      actionKey="portfolio.create"
      handleClose={handleClose}
      validate={(n) => validateName(n)}
      action={async (newName) =>
        newName ? create(newName) : Promise.resolve()
      }
    />
  );
};

export default CreateDialog;
