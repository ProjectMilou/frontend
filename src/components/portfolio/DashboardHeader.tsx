import React from 'react';
import { makeStyles } from '@material-ui/core';

const useStyles = makeStyles({
  header: {
    height: '350px',
    'background-color': '#122654',
  },
  text: {
    'font-size': '35pt',
    'color': '#EEF1FB',
    'padding-top': '5%',
    'padding-left' : '5%',
    width: '50%',
  },
});

const DashboardHeader: React.FC = () => {
  const classes = useStyles();
  return (
    <div className={classes.header}>
      <p className={classes.text}>
        Diversifying and balancing your portfolio helps you to optimize the
        ratio between risk and return.
      </p>
    </div>
  );
};

export default DashboardHeader;
