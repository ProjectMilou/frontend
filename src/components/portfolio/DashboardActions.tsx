import React from 'react';
import { useTranslation } from 'react-i18next';
import { IconButton, makeStyles, Tooltip } from '@material-ui/core';
import DeleteIcon from '@material-ui/icons/DeleteForever';
import DuplicateIcon from '@material-ui/icons/AddToPhotos';
import EditIcon from '@material-ui/icons/Edit';
import classNames from 'classnames';
import { PortfolioOverview } from '../../portfolio/APIClient';

const useStyles = makeStyles(() => ({
  action: { display: 'inline-block' },
  disabled: {
    cursor: 'not-allowed',
  },
}));

type DashboardActionsProps = {
  portfolio: PortfolioOverview;
  renamePortfolio: (id: string) => void;
  duplicatePortfolio: (id: string) => void;
  deletePortfolio: (id: string) => void;
};

const DashboardActions: React.FC<DashboardActionsProps> = ({
  portfolio,
  renamePortfolio,
  duplicatePortfolio,
  deletePortfolio,
}) => {
  const { t } = useTranslation();
  const classes = useStyles();

  return (
    <>
      <Tooltip title={t('portfolio.rename').toString()}>
        <div className={classes.action}>
          <IconButton
            onClick={(e) => {
              renamePortfolio(portfolio.id);
              e.stopPropagation();
            }}
          >
            <EditIcon />
          </IconButton>
        </div>
      </Tooltip>
      <Tooltip title={t('portfolio.duplicate').toString()}>
        <div className={classes.action}>
          <IconButton
            onClick={(e) => {
              duplicatePortfolio(portfolio.id);
              e.stopPropagation();
            }}
          >
            <DuplicateIcon />
          </IconButton>
        </div>
      </Tooltip>
      <Tooltip
        title={(portfolio.virtual
          ? t('portfolio.delete')
          : t('portfolio.deleteReal')
        ).toString()}
        // stop click event propagation here because the delete button can be disabled
        onClick={(e) => e.stopPropagation()}
      >
        <div
          className={classNames(
            classes.action,
            !portfolio.virtual && classes.disabled
          )}
        >
          <IconButton
            disabled={!portfolio.virtual}
            onClick={() => {
              deletePortfolio(portfolio.id);
            }}
          >
            <DeleteIcon />
          </IconButton>
        </div>
      </Tooltip>
    </>
  );
};

export default DashboardActions;
