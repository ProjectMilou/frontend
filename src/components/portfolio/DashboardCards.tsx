import React from 'react';
import { useTranslation } from 'react-i18next';
import {
  Card,
  CardActions,
  CardContent,
  GridList,
  GridListTile,
  IconButton,
  makeStyles,
  Tooltip,
  Typography,
} from '@material-ui/core';
import AddBoxIcon from '@material-ui/icons/AddBox';
import ImportIcon from '@material-ui/icons/Input';
import * as API from '../../portfolio/APIClient';
import { PortfolioOverview } from '../../portfolio/APIClient';
import DashboardActions from './DashboardActions';
import StyledNumberFormat from '../shared/StyledNumberFormat';
import { importPortfolio, portfolioDetails } from '../../portfolio/Router';

const useStyles = makeStyles(() => ({
  gridList: {
    width: '100%',
    height: 'auto',
    margin: '0 auto !important',
  },
  gridListTile: {
    minWidth: '12rem',
  },
  card: {
    cursor: 'pointer',
  },
  blankCard: {
    height: '13rem',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'transparent',
    borderStyle: 'dashed',
    borderColor: 'grey',
  },
  cardContent: {
    display: 'flex',
    flexDirection: 'column',
  },
  cardActions: {
    justifyContent: 'center',
  },
  portName: {
    fontSize: '1.5rem',
  },
  value: {
    marginTop: '1rem',
    alignSelf: 'center',
  },
}));

type CardComponentProps = {
  portfolio: PortfolioOverview;
  renamePortfolio: (id: string) => void;
  duplicatePortfolio: (id: string) => void;
  deletePortfolio: (id: string) => void;
};

const CardComponent: React.FC<CardComponentProps> = ({
  portfolio,
  renamePortfolio,
  duplicatePortfolio,
  deletePortfolio,
}) => {
  const classes = useStyles();
  const { t } = useTranslation();

  return (
    <Card className={classes.card}>
      <CardContent className={classes.cardContent}>
        <Typography className={classes.portName}>{portfolio.name}</Typography>
        <Typography color={portfolio.virtual ? 'textSecondary' : 'secondary'}>
          {portfolio.virtual ? t('portfolio.virtual') : t('portfolio.real')}
        </Typography>
        <div className={classes.value}>
          <StyledNumberFormat
            value={portfolio.value}
            suffix="â‚¬"
            size="24"
            fontWeight={400}
          />
        </div>
      </CardContent>
      <CardActions className={classes.cardActions}>
        <DashboardActions
          portfolio={portfolio}
          renamePortfolio={renamePortfolio}
          duplicatePortfolio={duplicatePortfolio}
          deletePortfolio={deletePortfolio}
        />
      </CardActions>
    </Card>
  );
};

export type DashboardCardsProps = {
  portfolios: API.PortfolioOverview[];
  renamePortfolio: (id: string) => void;
  duplicatePortfolio: (id: string) => void;
  deletePortfolio: (id: string) => void;
  createPortfolio: () => void;
};

const DashboardCards: React.FC<DashboardCardsProps> = ({
  portfolios,
  renamePortfolio,
  duplicatePortfolio,
  deletePortfolio,
  createPortfolio,
}) => {
  const classes = useStyles();
  const { t } = useTranslation();

  return (
    <GridList
      cellHeight="auto"
      cols={4}
      spacing={32}
      className={classes.gridList}
    >
      {portfolios.map((p) => (
        <GridListTile
          key={p.id}
          onClick={() => portfolioDetails(p.id)}
          className={classes.gridListTile}
        >
          <CardComponent
            key={p.id}
            portfolio={p}
            renamePortfolio={renamePortfolio}
            duplicatePortfolio={duplicatePortfolio}
            deletePortfolio={deletePortfolio}
          />
        </GridListTile>
      ))}
      <GridListTile
        className={classes.gridListTile}
        onClick={() => {
          createPortfolio();
        }}
      >
        <Card className={classes.blankCard}>
          <Tooltip title={t('portfolio.create').toString()}>
            <div>
              <IconButton>
                <AddBoxIcon fontSize="large" />
              </IconButton>
            </div>
          </Tooltip>
        </Card>
      </GridListTile>
      <GridListTile
        className={classes.gridListTile}
        onClick={() => importPortfolio()}
      >
        <Card className={classes.blankCard}>
          <Tooltip title={t('portfolio.dashboard.importPortfolio').toString()}>
            <div>
              <IconButton>
                <ImportIcon fontSize="large" />
              </IconButton>
            </div>
          </Tooltip>
        </Card>
      </GridListTile>
    </GridList>
  );
};

export default DashboardCards;
