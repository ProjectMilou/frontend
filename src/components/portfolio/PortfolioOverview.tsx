import React from 'react';
import { useTranslation } from 'react-i18next';
import { makeStyles, Typography } from '@material-ui/core';
import * as API from '../../portfolio/APIClient';
import DashboardViewSelect, {
  DashboardView,
} from '../shared/DashboardViewSelect';
import DashboardTable from './DashboardTable';
import DashboardCards from './DashboardCards';

export type PortfolioOverviewProps = {
  portfolios: API.PortfolioOverview[];
  selectPortfolio: (id: string) => void;
  renamePortfolio: (id: string) => void;
  duplicatePortfolio: (id: string) => void;
  deletePortfolio: (id: string) => void;
  createPortfolio: () => void;
};

const useStyles = makeStyles({
  header: { display: 'flex', marginBottom: '25px' },
  title: {
    fontSize: '48px',
    flex: '1',
    lineHeight: '72px',
    letterSpacing: '-0.015em',
  },
});

const PortfolioOverview: React.FC<PortfolioOverviewProps> = ({
  portfolios,
  selectPortfolio,
  renamePortfolio,
  duplicatePortfolio,
  deletePortfolio,
  createPortfolio,
}) => {
  const { t } = useTranslation();
  const classes = useStyles();
  const [view, setView] = React.useState<DashboardView>(DashboardView.Table);
  return (
    <>
      <div className={classes.header}>
        <div className={classes.title}>{t('portfolio.dashboard.title')}</div>
        {!!portfolios.length && (
          <DashboardViewSelect view={view} setView={setView} />
        )}
      </div>
      {view === DashboardView.Table ? (
        <DashboardTable
          portfolios={portfolios}
          selectPortfolio={selectPortfolio}
          renamePortfolio={renamePortfolio}
          duplicatePortfolio={duplicatePortfolio}
          deletePortfolio={deletePortfolio}
          createPortfolio={createPortfolio}
        />
      ) : (
        <DashboardCards
          portfolios={portfolios}
          selectPortfolio={selectPortfolio}
          renamePortfolio={renamePortfolio}
          duplicatePortfolio={duplicatePortfolio}
          deletePortfolio={deletePortfolio}
          createPortfolio={createPortfolio}
        />
      )}
    </>
  );
};

export default PortfolioOverview;
