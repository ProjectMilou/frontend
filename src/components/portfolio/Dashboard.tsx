import React, { Reducer } from 'react';
import { RouteComponentProps } from '@reach/router';
import { LinearProgress, makeStyles, Container } from '@material-ui/core';
import { useTranslation } from 'react-i18next';
import * as API from '../../portfolio/APIClient';
import ErrorMessage from '../shared/ErrorMessage';
import PortfolioOverview from './PortfolioOverview';
import DashboardHeader from '../shared/DashboardHeader';
import RenameDialog from './RenameDialog';
import DuplicateDialog from './DuplicateDialog';
import DeleteDialog from './DeleteDialog';
import CreateDialog from './CreateDialog';
import { isAuthenticationError } from '../../Errors';

export interface DashboardProps extends RouteComponentProps {
  token: string;
  selectPortfolio: (id: string) => void;
};

enum DialogType {
  None,
  CreatePortfolio,
  RenamePortfolio,
  DuplicatePortfolio,
  DeletePortfolio,
}

type Dialog = {
  type: DialogType;
  portfolioId?: string;
};

function portfolioName(
  portfolios?: API.PortfolioOverview[],
  id?: string
): string | undefined {
  return portfolios && id
    ? portfolios.find((p) => p.id === id)?.name
    : undefined;
}

/* A reducer is used to manage this component's state (portfolios, errors, open
 * dialog). The state should only be modified by dispatch calls, never directly.
 * Actions are not async, so API calls are made before dispatching.
 */

type State = {
  portfolios?: API.PortfolioOverview[];
  error?: Error;
  dialog: Dialog;
};

const initialState = { dialog: { type: DialogType.None } };

type SetPortfoliosAction = {
  type: 'setPortfolios';
  payload: API.PortfolioOverview[];
};
type CreateAction = {
  type: 'create';
  payload: { id: string; name: string };
};
type RenameAction = {
  type: 'rename';
  payload: { id: string; name: string };
};
type DuplicateAction = {
  type: 'duplicate';
  payload: { id: string; newId: string; newName: string };
};
type DeleteAction = {
  type: 'delete';
  payload: { id: string };
};
type SetErrorAction = {
  type: 'setError';
  payload: Error;
};
type ClearErrorAction = {
  type: 'clearError';
};
type OpenDialogAction = {
  type: 'openDialog';
  payload: Dialog;
};
type CloseDialogAction = {
  type: 'closeDialog';
};
type Actions =
  | SetPortfoliosAction
  | CreateAction
  | RenameAction
  | DuplicateAction
  | DeleteAction
  | SetErrorAction
  | ClearErrorAction
  | OpenDialogAction
  | CloseDialogAction;

function createPortfolioAction(
  portfolios: API.PortfolioOverview[] | undefined,
  { id, name }: CreateAction['payload']
) {
  return (
    portfolios && [
      ...portfolios,
      {
        id,
        name,
        virtual: true,
        positionCount: 0,
        value: 0,
        perf7d: 0,
        perf1y: 0,
        modified: new Date(),
      },
    ]
  );
}

function renamePortfolioAction(
  portfolios: API.PortfolioOverview[] | undefined,
  { id, name }: RenameAction['payload']
) {
  return (
    portfolios &&
    portfolios.map((p) => ({
      ...p,
      name: p.id === id ? name : p.name,
    }))
  );
}

function duplicatePortfolioAction(
  portfolios: API.PortfolioOverview[] | undefined,
  { id, newId, newName }: DuplicateAction['payload']
) {
  return (
    portfolios && [
      ...portfolios,
      {
        ...(portfolios.find((p) => p.id === id) as API.PortfolioOverview),
        id: newId,
        name: newName,
        virtual: true,
      },
    ]
  );
}

function deletePortfolioAction(
  portfolios: API.PortfolioOverview[] | undefined,
  { id }: DeleteAction['payload']
) {
  return portfolios && portfolios.filter((p) => p.id !== id);
}

function reducer(state: State, action: Actions) {
  switch (action.type) {
    case 'setPortfolios':
      return {
        ...state,
        portfolios: action.payload,
      };
    case 'create':
      return {
        ...state,
        portfolios: createPortfolioAction(state.portfolios, action.payload),
      };
    case 'rename':
      return {
        ...state,
        portfolios: renamePortfolioAction(state.portfolios, action.payload),
      };
    case 'duplicate':
      return {
        ...state,
        portfolios: duplicatePortfolioAction(state.portfolios, action.payload),
      };
    case 'delete':
      return {
        ...state,
        portfolios: deletePortfolioAction(state.portfolios, action.payload),
      };
    case 'setError':
      return {
        ...state,
        error: action.payload,
      };
    case 'clearError':
      return {
        ...state,
        error: undefined,
      };
    case 'openDialog':
      return {
        ...state,
        dialog: action.payload,
      };
    case 'closeDialog':
      return {
        ...state,
        dialog: initialState.dialog,
      };
    default:
      throw new Error();
  }
}

const useStyles = makeStyles({
  dashboard: {
    margin: '25px auto',
  },
});

const Dashboard: React.FC<DashboardProps> = ({ token, selectPortfolio }) => {
  const [state, dispatch] = React.useReducer<Reducer<State, Actions>>(
    reducer,
    initialState
  );

  const isMounted = React.useRef(true);
  const fetch = async () => {
    dispatch({ type: 'clearError' });
    try {
      const p = await API.list(token);
      if (isMounted.current) {
        dispatch({ type: 'setPortfolios', payload: p });
      }
    } catch (e) {
      if (isMounted.current) {
        dispatch({ type: 'setError', payload: e });
      }
    }
  };

  React.useEffect(() => {
    fetch();
    return () => {
      isMounted.current = false;
    };
    // deps must be empty because the function should only be called on mount.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const { t } = useTranslation();
  const classes = useStyles();

  return (
    <>
      <DashboardHeader>{t('portfolio.dashboard.headerText')}</DashboardHeader>
      {!state.portfolios && !state.error && (
        <div>
          <LinearProgress color="secondary" />
        </div>
      )}
      <Container maxWidth="lg" className={classes.dashboard}>
        {state.error && (
          <ErrorMessage
            error={state.error}
            messageKey="portfolio.dashboard.errorMessage"
            handling={
              isAuthenticationError(state.error)
                ? {
                  buttonText: 'error.action.login',
                  action: async () => {
                    // TODO: go back to login
                  },
                }
                : {
                  buttonText: 'error.action.retry',
                  action: fetch,
                }
            }
          />
        )}
        {state.portfolios && (
          <div>
            <PortfolioOverview
              portfolios={state.portfolios}
              selectPortfolio={selectPortfolio}
              renamePortfolio={(portfolioId) =>
                dispatch({
                  type: 'openDialog',
                  payload: { type: DialogType.RenamePortfolio, portfolioId },
                })
              }
              duplicatePortfolio={(portfolioId) =>
                dispatch({
                  type: 'openDialog',
                  payload: { type: DialogType.DuplicatePortfolio, portfolioId },
                })
              }
              deletePortfolio={(portfolioId) => {
                dispatch({
                  type: 'openDialog',
                  payload: { type: DialogType.DeletePortfolio, portfolioId },
                });
              }}
              createPortfolio={() => {
                dispatch({
                  type: 'openDialog',
                  payload: { type: DialogType.CreatePortfolio },
                });
              }}
            />
          </div>
        )}
      </Container>
      <CreateDialog
        open={state.dialog.type === DialogType.CreatePortfolio}
        handleClose={() => dispatch({ type: 'closeDialog' })}
        create={async (name) => {
          const id = await API.create(token, name);
          dispatch({ type: 'create', payload: { id, name } });
        }}
      />
      <RenameDialog
        initialName={portfolioName(state.portfolios, state.dialog.portfolioId)}
        open={state.dialog.type === DialogType.RenamePortfolio}
        handleClose={() => dispatch({ type: 'closeDialog' })}
        rename={async (name) => {
          if (state.dialog.portfolioId) {
            await API.rename(token, state.dialog.portfolioId, name);
            dispatch({
              type: 'rename',
              payload: { id: state.dialog.portfolioId, name },
            });
          }
        }}
      />
      <DuplicateDialog
        initialName={portfolioName(state.portfolios, state.dialog.portfolioId)}
        open={state.dialog.type === DialogType.DuplicatePortfolio}
        handleClose={() => dispatch({ type: 'closeDialog' })}
        duplicate={async (name) => {
          if (state.dialog.portfolioId) {
            const id = await API.duplicate(
              token,
              state.dialog.portfolioId,
              name
            );
            dispatch({
              type: 'duplicate',
              payload: {
                id: state.dialog.portfolioId,
                newId: id,
                newName: name,
              },
            });
          }
        }}
      />
      <DeleteDialog
        initialName={portfolioName(state.portfolios, state.dialog.portfolioId)}
        open={state.dialog.type === DialogType.DeletePortfolio}
        handleClose={() => dispatch({ type: 'closeDialog' })}
        deletePortfolio={async () => {
          if (state.dialog.portfolioId) {
            await API.deletePortfolio(token, state.dialog.portfolioId);
            dispatch({
              type: 'delete',
              payload: { id: state.dialog.portfolioId },
            });
          }
        }}
      />
    </>
  );
};

export default Dashboard;
