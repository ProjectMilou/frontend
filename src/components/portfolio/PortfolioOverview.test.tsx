import { render, fireEvent } from '@testing-library/react';
import * as React from 'react';
import PortfolioOverview, { PortfolioOverviewProps } from './PortfolioOverview';
import { MockOverview, MockOverviewTwo } from '../../portfolio/APIClient';

jest.mock('./DashboardTable', () => () => <table />);

describe('PortfolioOverview', () => {
  const defaultProps: PortfolioOverviewProps = {
    portfolios: [MockOverview, MockOverviewTwo],
    selectPortfolio: jest.fn(),
    renamePortfolio: jest.fn(),
    duplicatePortfolio: jest.fn(),
    deletePortfolio: jest.fn(),
  };

  const renderComponent = (newProps?: Partial<PortfolioOverviewProps>) => {
    const props = { ...defaultProps, ...newProps };
    return {
      ...render(<PortfolioOverview {...props} />),
      props,
    };
  };

  test('show message if no portfolios exist', () => {
    const { container, getByText } = renderComponent({ portfolios: [] });
    expect(getByText('portfolio.dashboard.title')).toBeInTheDocument();
    expect(getByText('portfolio.dashboard.noPortfolios')).toBeInTheDocument();
    // no table
    expect(container.querySelector('table')).toBeNull();
    // no view selection
    expect(container.querySelectorAll('button svg')).toHaveLength(0);
  });

  test('switch between table and cards view', () => {
    // TODO: Update test once cards view is implemented
    const { container, queryByText } = renderComponent();
    expect(queryByText('portfolio.dashboard.title')).toBeInTheDocument();
    expect(queryByText('portfolio.dashboard.noPortfolios')).toBeNull();
    expect(queryByText('cards')).toBeNull();
    expect(container.querySelectorAll('table')).toHaveLength(1);
    // buttons to change view
    const buttons = container.querySelectorAll('button svg');
    expect(buttons).toHaveLength(2);
    // change to cards view
    fireEvent.click(buttons[1]);
    expect(container.querySelectorAll('table')).toHaveLength(0);
    expect(queryByText('cards')).toBeInTheDocument();
    // change to table view
    fireEvent.click(buttons[0]);
    expect(container.querySelectorAll('table')).toHaveLength(1);
    expect(queryByText('cards')).toBeNull();
  });
});
