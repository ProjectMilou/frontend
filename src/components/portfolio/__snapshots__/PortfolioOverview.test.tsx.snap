// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PortfolioOverview show and handle error message 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-avatar"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle makeStyles-icon-2 MuiAvatar-colorDefault"
        >
          !
        </div>
      </div>
      <div
        class="MuiCardHeader-content"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5"
        >
          error.title.undefined
        </h2>
        <span
          class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          error.message.default
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body2"
      >
        error.message.undefined
      </p>
    </div>
  </div>
</div>
`;

exports[`PortfolioOverview show error without message that cannot be handled 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card-3 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-avatar"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle makeStyles-icon-4 MuiAvatar-colorDefault"
        >
          !
        </div>
      </div>
      <div
        class="MuiCardHeader-content"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5"
        >
          error.title.undefined
        </h2>
        <span
          class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          error.message.default
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body2"
      >
        error.message.undefined
      </p>
    </div>
  </div>
</div>
`;
